File: backend/pdgpuzzleapp/puzzle/scenarioFiller.py
import numpy as np
import random

from .mechanics import EMPTY, WALL, DOOR, ENEMY, BLOCK, PLAYER

def setRandomPlayer(dungeon):
    empty = np.argwhere(dungeon == EMPTY)
    player = random.choice(empty)
    dungeon[player[0],player[1]] = PLAYER
    return dungeon

def getAdjacentWalls(dungeon):
    # Obtener las posiciones de los 1s en la dungeon
    walls = np.argwhere(dungeon == WALL)
    adjacentWalls = []

    for i, j in walls:
        # Verificar si el 1 está adyacente a un 0 en forma de cruz
        if (i > 0 and dungeon[i-1][j] == EMPTY) or \
           (i < dungeon.shape[0]-1 and dungeon[i+1][j] == EMPTY) or \
           (j > 0 and dungeon[i][j-1] == 0) or \
           (j < dungeon.shape[1]-1 and dungeon[i][j+1] == EMPTY):
            adjacentWalls.append((i, j))

    return adjacentWalls

def setDoor(dungeon):
    allowDoorPosition = getAdjacentWalls(dungeon)
    door = random.choice(allowDoorPosition)
    dungeon[door[0],door[1]] = DOOR
    return dungeon

def setBlocksAndEnemys(dungeon,enemyFactor,blockFactor):
    empty = np.argwhere(dungeon == EMPTY)
    nEmpty = len(empty)
    nEnemys = nEmpty * enemyFactor
    nBlocks = nEmpty * blockFactor
    
    for _ in range(int(nEnemys)):
        enemy = random.choice(empty)
        dungeon[enemy[0],enemy[1]] = ENEMY
        empty = np.argwhere(dungeon == EMPTY)
    
    for _ in range(int(nBlocks)):
        block = random.choice(empty)
        dungeon[block[0],block[1]] = BLOCK
        empty = np.argwhere(dungeon == EMPTY)
    
    return dungeon

def getIntialSol(dungeon,enemyFactor,blockFactor):
    dungeon = setRandomPlayer(dungeon)
    dungeon = setDoor(dungeon)
    dungeon = setBlocksAndEnemys(dungeon,enemyFactor,blockFactor)
    
    return dungeon

def scenarioFiller(n, m, expansion_factor):
    n,m = int(n),int(m)
    """
    Improved generation of an n x m dungeon filled with zeros and ones ensuring all zeros are cross-connected.
    This function also includes the validation step using flood fill algorithm.

    Parameters:
    n (int): Number of rows.
    m (int): Number of columns.
    expansion_factor (float): Probability of a '1' expanding towards the center.

    Returns:
    np.ndarray: The filled dungeon if it is valid, otherwise it tries again.
    """

    def is_valid_matrix(dungeon):
        """
        Check if the dungeon is valid based on the flood fill algorithm.
        """
        n, m = dungeon.shape
        visited = np.zeros_like(dungeon, dtype=bool)

        def flood_fill(r, c):
            """
            Perform flood fill to mark connected zeros using an iterative approach with a stack.
            """
            # Stack to hold the cells to process
            stack = [(r, c)]

            # Loop until there are no more cells to process
            while stack:
                current_r, current_c = stack.pop()

                # Boundary and base case checks
                if current_r < 0 or current_r >= n or current_c < 0 or current_c >= m:
                    continue
                if visited[current_r, current_c] or dungeon[current_r, current_c] != 0:
                    continue

                # Mark as visited
                visited[current_r, current_c] = True

                # Add adjacent cells to stack
                stack.append((current_r + 1, current_c))  # down
                stack.append((current_r - 1, current_c))  # up
                stack.append((current_r, current_c + 1))  # right
                stack.append((current_r, current_c - 1))  # left

        # Start flood fill from the first zero found
        for i in range(n):
            for j in range(m):
                if dungeon[i, j] == EMPTY:
                    flood_fill(i, j)
                    break
            if dungeon[i, j] == EMPTY:
                break

        # If there are unvisited zeros, the dungeon is not valid
        for i in range(n):
            for j in range(m):
                if dungeon[i, j] == EMPTY and not visited[i, j]:
                    return False

        return True

    # Keep generating matrices until a valid one is found
    while True:

        # Create the initial dungeon with 0s
        dungeon = np.zeros((n, m))

        # Fill the contour with 1s
        dungeon[0, :] = WALL
        dungeon[:, 0] = WALL
        dungeon[-1, :] = WALL
        dungeon[:, -1] = WALL

        # Calculate the maximum expansion length (25% of the minimum dimension)
        max_expansion_length = min(n, m) // 4

        # Expansion process
        for _ in range(max_expansion_length+1):
            indices_to_expand = np.where(dungeon == WALL)
            for row, col in zip(*indices_to_expand):
                if np.random.rand() < expansion_factor:
                    if row > 0 and dungeon[row - 1, col] == EMPTY:
                        dungeon[row - 1, col] = 1
                    if row < n - 1 and dungeon[row + 1, col] == EMPTY:
                        dungeon[row + 1, col] = 1
                    if col > 0 and dungeon[row, col - 1] == EMPTY:
                        dungeon[row, col - 1] = 1
                    if col < m - 1 and dungeon[row, col + 1] == EMPTY:
                        dungeon[row, col + 1] = 1

        if is_valid_matrix(dungeon):
            return dungeon

File: backend/pdgpuzzle/puzzlea_pi/WebApplication1/Core/ScenarioFiller.cs
using System;
using System.Collections.Generic;
using System.Linq;
using WebApplication1.MechanicsClass;

namespace WebApplication1.ScenarioFille{
	public static class ScenarioFiller
{
	static Random random = new Random();

	public static int[,] SetRandomPlayer(int[,] dungeon)
	{
		var empty = GetPositions(dungeon, Mechanics.EMPTY);
		var player = empty[random.Next(empty.Count)];
		dungeon[player.Item1, player.Item2] = Mechanics.PLAYER;
		return dungeon;
	}

	public static List<(int, int)> GetAdjacentWalls(int[,] dungeon)
	{
		var walls = GetPositions(dungeon, Mechanics.WALL);
		var adjacentWalls = new List<(int, int)>();

		foreach (var (i, j) in walls)
		{
			if ((i > 0 && dungeon[i - 1, j] == Mechanics.EMPTY) ||
				(i < dungeon.GetLength(0) - 1 && dungeon[i + 1, j] == Mechanics.EMPTY) ||
				(j > 0 && dungeon[i, j - 1] == Mechanics.EMPTY) ||
				(j < dungeon.GetLength(1) - 1 && dungeon[i, j + 1] == Mechanics.EMPTY))
			{
				adjacentWalls.Add((i, j));
			}
		}

		return adjacentWalls;
	}

	public static int[,] SetDoor(int[,] dungeon)
	{
		var allowDoorPosition = GetAdjacentWalls(dungeon);
		var door = allowDoorPosition[random.Next(allowDoorPosition.Count)];
		dungeon[door.Item1, door.Item2] = Mechanics.DOOR;
		return dungeon;
	}

	public static int[,] SetBlocksAndEnemies(int[,] dungeon, double enemyFactor, double blockFactor)
	{
		var empty = GetPositions(dungeon, Mechanics.EMPTY);
		int nEmpty = empty.Count;
		int nEnemies = (int)(nEmpty * enemyFactor);
		int nBlocks = (int)(nEmpty * blockFactor);

		for (int i = 0; i < nEnemies; i++)
		{
			var enemy = empty[random.Next(empty.Count)];
			dungeon[enemy.Item1, enemy.Item2] = Mechanics.ENEMY;
			empty = GetPositions(dungeon, Mechanics.EMPTY);
		}

		for (int i = 0; i < nBlocks; i++)
		{
			var block = empty[random.Next(empty.Count)];
			dungeon[block.Item1, block.Item2] = Mechanics.BLOCK;
			empty = GetPositions(dungeon, Mechanics.EMPTY);
		}

		return dungeon;
	}

	public static int[,] GetInitialSolution(int[,] dungeon, double enemyFactor, double blockFactor)
	{
		dungeon = SetRandomPlayer(dungeon);
		dungeon = SetDoor(dungeon);
		dungeon = SetBlocksAndEnemies(dungeon, enemyFactor, blockFactor);
		return dungeon;
	}

	public static int[,] ScenarioFillerMethod(int n, int m, double expansionFactor)
	{
		while (true)
		{
			var dungeon = new int[n, m];

			// Fill the contour with walls
			for (int i = 0; i < n; i++)
			{
				dungeon[i, 0] = Mechanics.WALL;
				dungeon[i, m - 1] = Mechanics.WALL;
			}

			for (int j = 0; j < m; j++)
			{
				dungeon[0, j] = Mechanics.WALL;
				dungeon[n - 1, j] = Mechanics.WALL;
			}

			int maxExpansionLength = Math.Min(n, m) / 4;

			for (int step = 0; step <= maxExpansionLength; step++)
			{
				var indicesToExpand = GetPositions(dungeon, Mechanics.WALL);
				foreach (var (row, col) in indicesToExpand)
				{
					if (random.NextDouble() < expansionFactor)
					{
						if (row > 0 && dungeon[row - 1, col] == Mechanics.EMPTY)
						{
							dungeon[row - 1, col] = Mechanics.WALL;
						}
						if (row < n - 1 && dungeon[row + 1, col] == Mechanics.EMPTY)
						{
							dungeon[row + 1, col] = Mechanics.WALL;
						}
						if (col > 0 && dungeon[row, col - 1] == Mechanics.EMPTY)
						{
							dungeon[row, col - 1] = Mechanics.WALL;
						}
						if (col < m - 1 && dungeon[row, col + 1] == Mechanics.EMPTY)
						{
							dungeon[row, col + 1] = Mechanics.WALL;
						}
					}
				}
			}

			if (IsValidMatrix(dungeon))
			{
				return dungeon;
			}
		}
	}

	private static bool IsValidMatrix(int[,] dungeon)
	{
		int n = dungeon.GetLength(0);
		int m = dungeon.GetLength(1);
		var visited = new bool[n, m];

		void FloodFill(int r, int c)
		{
			var stack = new Stack<(int, int)>();
			stack.Push((r, c));

			while (stack.Count > 0)
			{
				var (currentR, currentC) = stack.Pop();

				if (currentR < 0 || currentR >= n || currentC < 0 || currentC >= m)
					continue;
				if (visited[currentR, currentC] || dungeon[currentR, currentC] != Mechanics.EMPTY)
					continue;

				visited[currentR, currentC] = true;

				stack.Push((currentR + 1, currentC)); // down
				stack.Push((currentR - 1, currentC)); // up
				stack.Push((currentR, currentC + 1)); // right
				stack.Push((currentR, currentC - 1)); // left
			}
		}

		for (int i = 0; i < n; i++)
		{
			for (int j = 0; j < m; j++)
			{
				if (dungeon[i, j] == Mechanics.EMPTY)
				{
					FloodFill(i, j);
					break;
				}
			}
		}

		for (int i = 0; i < n; i++)
		{
			for (int j = 0; j < m; j++)
			{
				if (dungeon[i, j] == Mechanics.EMPTY && !visited[i, j])
				{
					return false;
				}
			}
		}

		return true;
	}

	private static List<(int, int)> GetPositions(int[,] dungeon, int value)
	{
		var positions = new List<(int, int)>();
		for (int i = 0; i < dungeon.GetLength(0); i++)
		{
			for (int j = 0; j < dungeon.GetLength(1); j++)
			{
				if (dungeon[i, j] == value)
				{
					positions.Add((i, j));
				}
			}
		}
		return positions;
	}
}
}
